v2/writer.go

```diff
@@ -0,0 +1,86 @@
+package car
+
+import (
+	"bytes"
+	"context"
+	"github.com/ipfs/go-cid"
+	format "github.com/ipfs/go-ipld-format"
+	car_v1 "github.com/ipld/go-car"
+	"github.com/willscott/carbs"
+	"io"
+)
+
+type (
+	// Writer writes CAR v2 into a give io.Writer.
+	Writer struct {
+		Walk         car_v1.WalkFunc
+		IndexCodec   carbs.IndexCodec
+		NodeGetter   format.NodeGetter
+		IndexPadding uint64
+		w            io.Writer
+	}
+)
+
+// NewWriter instantiates a new CAR v2 writer.
+// The writer instantiated uses `carbs.IndexSorted` as the index codec,
+// and `car_v1.DefaultWalkFunc` as the default walk function.
+func NewWriter(ng format.NodeGetter, w io.Writer) *Writer {
+	return &Writer{
+		Walk:       car_v1.DefaultWalkFunc,
+		IndexCodec: carbs.IndexSorted,
+		NodeGetter: ng,
+		w:          w,
+	}
+}
+
+// Write writes the given root CIDs according to CAR v2 specification, traversing the DAG using the
+// Writer#Walk function.
+func (w *Writer) Write(ctx context.Context, roots []cid.Cid) (err error) {
+	err = w.writePrefix()
+	if err != nil {
+		return
+	}
+
+	buf, err := w.encodeCarV1(ctx, roots)
```

Comment:
I think a bit of docs as to why we're buffering the entirety of the inner carv1 would be nice. it seems like the only reason is building the index.

Reply:
Well spotted. will do. 

For reference, like you said we need to read the entire CAR v1 to then generate the index. Another less important reason is to set the size in CAR v2 `Header`; with a `Write` API that takes root CIDs we need to generate the CAR v1 to know its size. As far as I can see CAR v1 APIs don't allow the size of a CAR to be calculated without encoding it.

Reply:
If you just needed the size, you could still figure that out without buffering the entire thing thanks to the io.Writer interface. You could layer an io.Writer on top of `w` when encoding the carv1, which would simply forward the `Write` calls while also keeping an incremental count of the `n int` returned by each Write call, e.g. in a uint64. This is how you typically count the total or "current progress" amount of bytes that go through an io.Reader or io.Writer withotu buffering the whole thing, anyway.

Reply:
but the size needs to be written before all the car bytes.

For efficiency, we should allow writing to a `WriterAt` where we can seek back and fill in the header after the carv1 gets streamed.

Reply:
aha great suggestion thank you